src/main/java/io/cdap/plugin/sap/SapODataSource.java

```diff
@@ -129,86 +152,322 @@ public Schema getSchema() {
+      return entityType.getProperties().stream()
+        .filter(p -> config.getSelectProperties().isEmpty() || config.getSelectProperties().contains(p.getName()))
+        .collect(HashMap::new, (m, v) -> m.put(v.getName(), v.getAnnotations()), HashMap::putAll);
+    } catch (ODataException e) {
+      throw new InvalidStageException("Unable to get metadata annotations for the entity type: " + e.getMessage(), e);
+    }
+  }
+
+  private Schema.Field getSchemaField(PropertyMetadata propertyMetadata, boolean includeAnnotations) {
+    Schema nonNullableSchema = propertyToSchema(propertyMetadata);
     Schema schema = propertyMetadata.isNullable() ? Schema.nullableOf(nonNullableSchema) : nonNullableSchema;
-    return Schema.Field.of(propertyMetadata.getName(), schema);
+    List<ODataAnnotation> annotations = propertyMetadata.getAnnotations();
+
+    return includeAnnotations && annotations != null && !annotations.isEmpty()
+      ? getFieldWithAnnotations(propertyMetadata, schema) : Schema.Field.of(propertyMetadata.getName(), schema);
+  }
+
+  private Schema.Field getFieldWithAnnotations(PropertyMetadata propertyMetadata, Schema valueSchema) {
+    String propertyName = propertyMetadata.getName();
+    List<ODataAnnotation> annotations = propertyMetadata.getAnnotations();
+    List<Schema.Field> fields = annotations.stream()
+      .map(annotation -> Schema.Field.of(annotation.getName(), annotationToFieldSchema(propertyName, annotation)))
+      .collect(Collectors.toList());
+    Schema.Field metadataField = Schema.Field.of(SapODataConstants.METADATA_ANNOTATIONS_FIELD_NAME,
+                                                 Schema.recordOf(propertyName + "-metadata-record", fields));
+
+    Schema valueWithMetadataSchema = Schema.recordOf(propertyName + "-value-with-metadata-record",
+                                                     Schema.Field.of(SapODataConstants.VALUE_FIELD_NAME, valueSchema),
+                                                     metadataField);
+    return Schema.Field.of(propertyName, valueWithMetadataSchema);
+  }
+
+  private Schema annotationToFieldSchema(String fieldName, ODataAnnotation oDataAnnotation) {
+    if (oDataAnnotation instanceof OData2Annotation) {
+      // OData 2 annotations are strings
+      return Schema.of(Schema.Type.STRING);
+    }
+
+    // OData 4 annotations
+    OData4Annotation annotation = (OData4Annotation) oDataAnnotation;
+    CsdlExpression expression = annotation.getExpression();
+    Schema expressionSchema = expressionToFieldSchema(fieldName, expression);
+    Schema annotationsSchema = nestedAnnotationsSchema(annotation.getName(), annotation.getAnnotations());
+    String recordName = fieldName + "-" + annotation.getName();
+    return SapODataConstants.Annotation.schema(recordName, expressionSchema, annotationsSchema);
+  }
+
+  private @Nullable Schema nestedAnnotationsSchema(String annotationName, Map<String, OData4Annotation> annotations) {
```

Comment:
style: nullable annotations go on top of the method

```
@Nullable
private Schema ...
```

Reply:
Fixed.